src/types/orchestrationTypes.ts

```diff
@@ -1,13 +1,18 @@
 import { FunctionOptions, FunctionTrigger, InvocationContext } from "@azure/functions";
-import { TokenSource } from ".";
+import { Task, TokenSource } from ".";
 import { DurableOrchestrationContext } from "../classes";
 
 /**
  * Type of a Generator that can be registered as an orchestration
+ * @param T the return value of the orchestration
  */
-export type OrchestrationHandler = (
+export type OrchestrationHandler<T = any> = (
     context: OrchestrationContext
-) => Generator<unknown, unknown, any>;
+) => Generator<
+    Task, // orchestrations yield Task types
+    T, // return type of the orchestration
+    any // what the SDK passes back to the orchestration
```

Comment:
&lt;continuing the TNext discussion on a separate thread&gt;

> `TNext` is the result of the Durable APIs, which don't always have a consistent type, that's why I left it as `any` here. But again, it's determined by the SDK, not user code. Specifying `TNext` as, for example, `string`, means I will write an orchestration that only calls Durable APIs that return a string, which doesn't seem like a particularly common use-case, as opposed to specifying the return type.

Can you give a few of the most common examples of durable apis and what the type of TNext would be?

Reply:
The simplest example would be the `context.df.callActivity()` API, in which case the `TNext` type would be whatever was the return type of the Activity, so that API alone could be anything, and could vary from one call to the next.

```TS
const orchestrator: OrchestrationHandler = function* (context: OrchestrationContext) {
    yield context.df.callActivity('returnsString'); // type of TNext here is string
    yield context.df.callActivity('returnsNumber'); // type of TNext here is number
    yield context.df.callHttp({ method: 'get', url: 'http://someApi' }); // type of TNext here is HttpResponse
};

df.app.activity('returnsString', {
    handler: (): string => {
        return 'string';
    },
});

df.app.activity('returnsNumber', {
    handler: (): number => {
        return 1;
    },
});
```

Reply:
But I want to be clearer, that it's SDK code, not user code, that is affected by the type of `TNext`, because it's SDK code that calls `.next()` on the generator in this line here: https://github.com/Azure/azure-functions-durable-js/blob/4b4eae8a5dc9b1571be6cdfda229b7ae2d0fba4a/src/taskorchestrationexecutor.ts#L388-L390

So `TNext` specifies what type `taskValue` is allowed to be in the `this.generator.next(taskValue)` line, and it can really be anything.

Reply:
Yep, that's right. For us `TNext` represents an external result (over the network/an external Function invocation), so it would be peculiar if a user restricted themselves to only receiving a specific type here. The type `any` makes sense here because we really can't control / type-check what comes over the network.

Reply:
For the sake of reference, here's the orchestration handlers as of today:
```TS
const orchestrator: OrchestrationHandler = function* (context: OrchestrationContext) {
    const outputs = [];
    outputs.push(yield context.df.callActivity(activityName, 'Tokyo'));
    outputs.push(yield context.df.callActivity(activityName, 'Seattle'));
    outputs.push(yield context.df.callActivity(activityName, 'Cairo'));

    return outputs;
};

const helloActivity: ActivityHandler = (input: string): string => {
    return `Hello, ${input}`;
};
```
Our own orchestration template is only yielding `callActivity` on `helloActivity` which is the user's own code and always returns a string. Doesn't that mean `string` would be a logical value for `TNext` in this case? The other example Hossam gave me was `callHttp`, but even in that case we know the type would always be `HttpResponse`

It's not just about `TNext`. Hossam suggested we specify `TReturn` as `string[]` for the template, but that doesn't do anything (in terms of build errors, preventing bugs, etc.) unless we also change the type of `outputs` which is currently `any[]`. And what does the type of `outputs` depend on? `TNext`. I think `TReturn` and `TNext` are inextricably linked and I think the Node.js designers probably saw that when designing the type.